Implement a unique user identifier across the system

ashensw opened the pull request:
## Purpose
> This is to introduce a unique user identifier across the system which resolves https://github.com/wso2/product-is/issues/6753

## Goals
> This PR will introduce a user store manager interface that consists of a new set of user management APIs which work with unique user ID. 

## Approach
> This feature can be disabled or enabled via the following configuration in `carbon.xml`. 
`<UniqueUserID>
        <Enabled>true</Enabled>
    </UniqueUserID>`

## User stories
> Summary of user stories addressed by this change>

## Release note
> Brief description of the new feature or bug fix as it will appear in the release notes

## Documentation
> Link(s) to product documentation that addresses the changes of this PR. If no doc impact, enter “N/A” plus brief explanation of why there’s no doc impact

## Training
> Link to the PR for changes to the training content in https://github.com/wso2/WSO2-Training, if applicable

## Certification
> Type “Sent” when you have provided new/updated certification questions, plus four answers for each question (correct answer highlighted in bold), based on this change. Certification questions/answers should be sent to certification@wso2.com and NOT pasted in this PR. If there is no impact on certification exams, type “N/A” and explain why.

## Marketing
> Link to drafts of marketing content that will describe and promote this feature, including product page changes, technical articles, blog posts, videos, etc., if applicable

## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

## Samples
> Provide high-level details about the samples related to this feature

## Related PRs
> List any other related PRs

## Migrations (if applicable)
> Describe migration steps and platforms on which migration has been tested

## Test environment
> List all JDK versions, operating systems, databases, and browser/versions on which this feature/fix was tested
 
## Learning
> Describe the research phase and any blog posts, patterns, libraries, or add-ons you used to solve the problem.

ashensw committed:
Add unique user ID feature

ashensw committed:
Change add user API to return the user

ashensw committed:
Refactor existing interfaces

ashensw committed:
Add NotImplementedException class

ashensw committed:
Refactor methods

ashensw committed:
Fix test failures

ashensw committed:
Refactor the code

ashensw committed:
Introduce addRoleWithID method

ashensw committed:
Refactor the code to preserve java8 compatibility

ashensw committed:
Refactor the code to fix test failures

ashensw committed:
Add UniqueUserID config to carbon.xml

ashensw committed:
Refactor the code

ashensw committed:
Make unique ID generation logic extensible

ashensw committed:
Refactor the code

ashensw committed:
Refactor the code

denuwanthi commented on core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/UniqueIDUserStoreManager.java:
```diff
@@ -0,0 +1,305 @@
+
+/**
+ * The interface to read data from a user store.
+ * <p/>
```
unnecessary character?

ashensw replied:
Fixed and updated the PR.

madurangasiriwardena commented on core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/NotImplementedException.java:
```diff
@@ -0,0 +1,46 @@
+/*
+ *
```
Unnecessary empty line. Fix for all the places.

ashensw replied:
Fixed and updated the PR.

madurangasiriwardena commented on core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/NotImplementedException.java:
```diff
@@ -0,0 +1,46 @@
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ *
```
Unnecessary empty line. Fix for all the places.

ashensw replied:
Fixed and updated the PR.

madurangasiriwardena commented on core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/common/AbstractUserStoreManager.java:
```diff
@@ -207,6 +213,25 @@ public Object run() throws Exception {
                                                                  String[] propertyNames, String profileName) throws UserStoreException;
+
+    /**
+     * This method is used by the support system to read properties.
```
Add a more meaningful doc comment.

ashensw replied:
Fixed and updated the PR.

madurangasiriwardena commented on core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/common/AbstractUserStoreManager.java:
```diff
@@ -207,6 +213,25 @@ public Object run() throws Exception {
+     * @return user properties of the given user.
+     * @throws UserStoreException Thrown by the underlying UserStoreManager.
+     */
+    protected Map<String, String> getUserPropertyValuesWithID(String userID, String[] propertyNames, String profileName)
```
shouldn't this be doGetUserPropertyValuesWithID

madurangasiriwardena commented on core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/common/AbstractUserOperationEventListener.java:
```diff
@@ -160,7 +156,7 @@ public boolean doPostDeleteUserClaimValue(String userName, UserStoreManager user
     public boolean doPreAddRole(String roleName, String[] userList, Permission[] permissions,
-                                UserStoreManager userStoreManager) throws UserStoreException {
+            UserStoreManager userStoreManager) throws UserStoreException {
         return true;
```
new line?

ashensw replied:
Fixed and updated the PR.

madurangasiriwardena commented on core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/common/AbstractUserOperationEventListener.java:
```diff
@@ -175,7 +171,37 @@ public boolean doPreAddRole(String roleName, String[] userList, Permission[] per
     public boolean doPostAddRole(String roleName, String[] userList, Permission[] permissions,
-                                 UserStoreManager userStoreManager) throws UserStoreException {
+            UserStoreManager userStoreManager) throws UserStoreException {
+        return true;
```
new line

ashensw replied:
Fixed and updated the PR.

madurangasiriwardena commented on core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/common/AbstractUserOperationEventListener.java:
```diff
@@ -289,9 +308,278 @@ public boolean doPreUpdateRoleListOfUser(String userName, String[] deletedRoles,
-                                              String[] newRoles, UserStoreManager userStoreManager)
+    public boolean doPostUpdateRoleListOfUser(String userName, String[] deletedRoles, String[] newRoles,
+            UserStoreManager userStoreManager) throws UserStoreException {
+        return true;
```
new line

ashensw replied:
Fixed and updated the PR.

madurangasiriwardena commented on core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/common/AbstractUserOperationEventListener.java:
```diff
@@ -289,9 +308,278 @@ public boolean doPreUpdateRoleListOfUser(String userName, String[] deletedRoles,
+    @Override
+    public boolean doPreGetUserListWithID(String claimUri, String claimValue, final List<User> returnUsersList,
+            UserStoreManager userStoreManager) throws UserStoreException {
+        return true;
```
new line

ashensw replied:
Fixed and updated the PR.

madurangasiriwardena commented on core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/common/AbstractUserOperationEventListener.java:
```diff
@@ -289,9 +308,278 @@ public boolean doPreUpdateRoleListOfUser(String userName, String[] deletedRoles,
+    @Override
+    public boolean doPostGetPaginatedUserListWithID(String claimUri, String claimValue, final List<User> returnValues,
+            UserStoreManager userStoreManager) throws UserStoreException {
+        return true;
```
new line

ashensw replied:
Fixed and updated the PR.

madurangasiriwardena commented on core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/common/AbstractUserStoreManager.java:
```diff
@@ -254,6 +295,25 @@ public Object run() throws Exception {
             throws UserStoreException;
 
+    /**
+     * Given the preferred user name and a credential object, the implementation code must
```
isn't it "preferred user name property"

ashensw replied:
preferred user name consist of both preferred username property and value.

IsuraD replied:
is this a claim URI?

madurangasiriwardena commented on core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/common/AbstractUserStoreManager.java:
```diff
@@ -269,6 +329,26 @@ protected abstract void  doAddUser(String userName, Object credential, String[]
+     * @param requirePasswordChange whether password required is need.
+     * @throws UserStoreException An unexpected exception has occurred.
+     */
+    protected User doAddUserWithID(String userName, Object credential, String[] roleList, Map<String, String> claims,
```
is this userName or userId?

ashensw replied:
It's userName. When adding the user, the username should be provided with the credentials. So the UUID will be generated inside the method.

madurangasiriwardena commented on core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/common/AbstractUserStoreManager.java:
```diff
@@ -386,12 +623,47 @@ protected abstract void doUpdateRoleListOfUser(String userName, String[] deleted
     }
 
+    /**
+     * Only gets the internal roles of the user with internal domain name.
```
Is this returning the "INTERNAL" domain only or will it return the "APPLICATION" role as well?

madurangasiriwardena commented on core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/common/AbstractUserStoreManager.java:
```diff
@@ -386,12 +623,47 @@ protected abstract void doUpdateRoleListOfUser(String userName, String[] deleted
+     * @throws UserStoreException An unexpected exception has occurred.
+     */
+    protected String[] doGetInternalRoleListOfUserWithID(String userID, String filter) throws UserStoreException {
+        if (Boolean.parseBoolean(realmConfig.getUserStoreProperty(MULIPLE_ATTRIBUTE_ENABLE))) {
```
new line

ashensw replied:
Fixed and updated the PR.

madurangasiriwardena commented on core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/common/AbstractUserStoreManager.java:
```diff
@@ -386,12 +623,47 @@ protected abstract void doUpdateRoleListOfUser(String userName, String[] deleted
+    protected String[] doGetInternalRoleListOfUserWithID(String userID, String filter) throws UserStoreException {
+        if (Boolean.parseBoolean(realmConfig.getUserStoreProperty(MULIPLE_ATTRIBUTE_ENABLE))) {
+            String userNameAttribute = realmConfig.getUserStoreProperty(LDAPConstants.USER_NAME_ATTRIBUTE);
+            if (userNameAttribute != null && userNameAttribute.trim().length() > 0) {
```
Can we use StringUtils.isEmpty ?

ashensw replied:
Yes. Fixed and updated the PR.

madurangasiriwardena commented on core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/common/AbstractUserStoreManager.java:
```diff
@@ -386,12 +623,47 @@ protected abstract void doUpdateRoleListOfUser(String userName, String[] deleted
+            String userNameAttribute = realmConfig.getUserStoreProperty(LDAPConstants.USER_NAME_ATTRIBUTE);
+            if (userNameAttribute != null && userNameAttribute.trim().length() > 0) {
+                Map<String, String> map;
+                map = getUserPropertyValuesWithID(userID, new String[] { userNameAttribute }, null);
```
Merge the declaration and initialization

ashensw replied:
Fixed and updated the PR.

ashensw committed:
Add getUserWithID API

ashensw committed:
Fix review changes

ashensw committed:
Add getUserIdWithID API

ashensw committed:
Fix startup errors

madurangasiriwardena commented on core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/jdbc/JDBCUserStoreManager.java:
```diff
@@ -465,6 +468,130 @@ private DataSource getJDBCDataSource() throws UserStoreException {
+        try {
+            givenMax = Integer
+                    .parseInt(realmConfig.getUserStoreProperty(UserCoreConstants.RealmConfig.PROPERTY_MAX_USER_LIST));
+        } catch (Exception e) {
```
can we narrow this down to NumberFormatException

ashensw replied:
Yes. will do.

madurangasiriwardena commented on core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/jdbc/JDBCUserStoreManager.java:
```diff
@@ -465,6 +468,130 @@ private DataSource getJDBCDataSource() throws UserStoreException {
+        try {
+            searchTime = Integer
+                    .parseInt(realmConfig.getUserStoreProperty(UserCoreConstants.RealmConfig.PROPERTY_MAX_SEARCH_TIME));
+        } catch (Exception e) {
```
can we narrow this down to NumberFormatException

ashensw replied:
Yes. will do.

madurangasiriwardena commented on core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/jdbc/JDBCUserStoreManager.java:
```diff
@@ -465,6 +468,130 @@ private DataSource getJDBCDataSource() throws UserStoreException {
+                prepStmt.setQueryTimeout(searchTime);
+            } catch (Exception e) {
+                // this can be ignored since timeout method is not implemented
+                log.debug(e);
```
Let's add a condition to check if the debug logs enabled and add more information to the log.

madurangasiriwardena commented on core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/jdbc/JDBCUserStoreManager.java:
```diff
@@ -465,6 +468,130 @@ private DataSource getJDBCDataSource() throws UserStoreException {
+
+            dbConnection = getDBConnection();
+
+            if (dbConnection == null) {
```
Seems to be there is no possibility of returning null from getDBConnection(); Check and confirm.
If there is a possibility, add more information to the exception.

madurangasiriwardena commented on core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/jdbc/JDBCUserStoreManager.java:
```diff
@@ -465,6 +468,130 @@ private DataSource getJDBCDataSource() throws UserStoreException {
+
+            List<User> userList = new ArrayList<>();
+
+            dbConnection = getDBConnection();
```
use try with resource.

madurangasiriwardena commented on core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/jdbc/JDBCUserStoreManager.java:
```diff
@@ -465,6 +468,130 @@ private DataSource getJDBCDataSource() throws UserStoreException {
+
+            sqlStmt = realmConfig.getUserStoreProperty(JDBCRealmConstants.GET_USER_ID_FILTER);
+
+            prepStmt = dbConnection.prepareStatement(sqlStmt);
```
use try with resource.

madurangasiriwardena commented on core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/jdbc/JDBCUserStoreManager.java:
```diff
@@ -465,6 +468,130 @@ private DataSource getJDBCDataSource() throws UserStoreException {
+            }
+
+            try {
+                rs = prepStmt.executeQuery();
```
use try with resource.

madurangasiriwardena commented on core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/jdbc/JDBCUserStoreManager.java:
```diff
@@ -1268,6 +1499,50 @@ public boolean doCheckExistingUser(String userName) throws UserStoreException {
+
+        if (userIDs.length > 1) {
+            throw new UserStoreException(
+                    "Invalid scenario. Multiple users cannot be found for the given user attribute.");
```
Let's change the error message. "Multiple users found with the given value xyz for the claim abc"

madurangasiriwardena commented on core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/jdbc/JDBCUserStoreManager.java:
```diff
@@ -1362,6 +1637,133 @@ public boolean doAuthenticate(String userName, Object credential) throws UserSto
+            return null;
+        }
+
+        // add the properties
```
This comment seems to be not explaining what you are doing.

madurangasiriwardena commented on core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/jdbc/JDBCUserStoreManager.java:
```diff
@@ -1362,6 +1637,133 @@ public boolean doAuthenticate(String userName, Object credential) throws UserSto
+        boolean isAuthed = false;
+
+        try {
+            dbConnection = getDBConnection();
```
Use try with resource for connections, prepared statements and result sets, for all the new codes you are adding.

madurangasiriwardena commented on core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/jdbc/JDBCUserStoreManager.java:
```diff
@@ -1362,6 +1637,133 @@ public boolean doAuthenticate(String userName, Object credential) throws UserSto
+                                    CarbonContext.getThreadLocalCarbonContext().getUserRealm()
+                                            .getRealmConfiguration()));
+                        } catch (org.wso2.carbon.user.api.UserStoreException e) {
+                            throw new UserStoreException(e);
```
Either throw the new error with more information or don't cat this error. Do the same thing in all the similar places.

ashensw committed:
Add UniqueIDJDBCUserStoreManager

ashensw committed:
Add PaginatedUserStoreManager APIs

JKAUSHALYA committed:
Add unique user id manager.

Added the unique user id manger to maintain unique id in the abstract level.

JKAUSHALYA committed:
Add the AuthenticationResult to the authenticate method.

madurangasiriwardena commented on core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/UserCoreConstants.java:
```diff
@@ -220,6 +221,7 @@
         public static final String DOMAIN_NAME_XPATH = "//UserStoreManager/Property[@name='DomainName']";
         public static final String LEADING_OR_TRAILING_SPACE_ALLOWED_IN_USERNAME =
                 "LeadingOrTrailingSpaceAllowedInUserName";
+        public static final String IS_USER_ID_ENABLED = "IsUserIDEnabled";
```
Would it be better to rename this config to "UserIDEnabled"?

ashensw committed:
Refactor the code

IsuraD commented on core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/NotImplementedException.java:
```diff
@@ -0,0 +1,44 @@
+    public NotImplementedException(Throwable cause) {
+        super(cause);
+    }
+
```
unncessary new line

IsuraD commented on core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/UniqueIDUserStoreManager.java:
```diff
@@ -0,0 +1,401 @@
+     * @return The auto increment user ID stored in underlying JDBC user store.
+     * @throws UserStoreException Thrown by the underlying UserStoreManager.
+     */
+    int getUserIdWithID(String userID) throws UserStoreException;
```
Why do we need this?

ashensw replied:
Removed and updated the PR

IsuraD commented on core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/UniqueIDUserStoreManager.java:
```diff
@@ -0,0 +1,401 @@
+     * @return added user.
+     * @throws UserStoreException Thrown by the underlying UserStoreManager.
+     */
+    User addUserWithID(String userName, Object credential, String[] roleList, Map<String, String> claims,
```
This is not needed

ashensw replied:
Removed and updated the PR

IsuraD commented on core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/UniqueIDUserStoreManager.java:
```diff
@@ -0,0 +1,401 @@
+     * @return tenant ID.
+     * @throws UserStoreException Thrown by the underlying UserStoreManager.
+     */
+    int getTenantIdWithID(String userID) throws UserStoreException;
```
This is not needed. Let's remove this.

ashensw replied:
Removed and updated the PR

IsuraD commented on core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/UniqueIDUserStoreManager.java:
```diff
@@ -0,0 +1,401 @@
+     * @param permissions The permissions of the role.
+     * @throws UserStoreException Thrown by the underlying UserStoreManager.
+     */
+    void addRoleWithID(String roleName, String[] userIDList, Permission[] permissions) throws UserStoreException;
```
Remove this.

IsuraD commented on core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/UniqueIDUserStoreManager.java:
```diff
@@ -0,0 +1,401 @@
+     *
+     * @param roleName     role name.
+     * @param deletedUsers deleted users.
+     * @param newUsers     new users.
```
@param newUsers     new user Ids.

ashensw replied:
made the changes and updated the PR

ashensw committed:
Fix code review changes

ashensw committed:
Refator the code

ashensw committed:
Refactor the code

JKAUSHALYA committed:
Add unique user ID capability to new abstract methods.

IsuraD commented on core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/UserCoreConstants.java:
```diff
@@ -154,6 +154,7 @@
         public static final String PROPERTY_MAX_USER_LIST = "MaxUserNameListLength";
         public static final String PROPERTY_MAX_ROLE_LIST = "MaxRoleNameListLength";
         public static final String PROPERTY_MAX_SEARCH_TIME = "MaxSearchQueryTime";
+        public static final String PROPERTY_MAX_USER_LIST_FOR_SCIM = "EnableMaxUserLimitForSCIM";
```
Why SCIM specific limit in user core?

IsuraD commented on core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/common/AbstractUserStoreManager.java:
```diff
@@ -392,6 +667,35 @@ protected abstract void doUpdateRoleListOfUser(String userName, String[] deleted
+
+        if (Boolean.parseBoolean(realmConfig.getUserStoreProperty(MULIPLE_ATTRIBUTE_ENABLE))) {
+            String userNameAttribute = realmConfig.getUserStoreProperty(LDAPConstants.USER_NAME_ATTRIBUTE);
+            if (StringUtils.isEmpty(userNameAttribute)) {
```
should be isNotEmpty

IsuraD commented on core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/common/AbstractUserStoreManager.java:
```diff
@@ -392,6 +667,35 @@ protected abstract void doUpdateRoleListOfUser(String userName, String[] deleted
+     */
+    protected String[] doGetInternalRoleListOfUserWithID(String userID, String filter) throws UserStoreException {
+
+        if (Boolean.parseBoolean(realmConfig.getUserStoreProperty(MULIPLE_ATTRIBUTE_ENABLE))) {
```
Since this is the userID, we don't need to consider abound Multi-attribute enable property?

IsuraD commented on core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/common/AbstractUserStoreManager.java:
```diff
@@ -418,9 +722,43 @@ protected abstract void doUpdateRoleListOfUser(String userName, String[] deleted
+    protected Map<String, List<String>> doGetInternalRoleListOfUsersWithID(List<String> userIDs, String domainName)
+            throws UserStoreException {
+
+        if (Boolean.parseBoolean(realmConfig.getUserStoreProperty(MULIPLE_ATTRIBUTE_ENABLE))) {
```
input is userIds. Do we need to consider  MULIPLE_ATTRIBUTE_ENABLE property?

IsuraD commented on core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/common/AbstractUserStoreManager.java:
```diff
@@ -701,9 +1154,13 @@ private boolean authenticateInternal(String userName, Object credential, boolean
             try {
                 // Let's authenticate with the primary UserStoreManager.
+                // Get the relevant userID for the given username.
+                if (isUniqueUserIdEnabled()) {
```
we have to check the  isUniqueUserIdEnabled property in abstractUserStoreManager instance. Let's this approach is correct or not.

IsuraD commented on core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/common/AbstractUserStoreManager.java:
```diff
@@ -1337,18 +2089,81 @@ private void handlePostListPaginatedUsers(String filter, int limit, int offset,
+        String[] userIDs = getUserListInternal(claim, claimValue, profileName);
+        List<User> users = new ArrayList<>();
+        for (String userID : userIDs) {
+            User user = new User(userID);
```
Is this return only ids? Don't we need to populate user object.

IsuraD commented on core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/common/AbstractUserStoreManager.java:
```diff
@@ -1366,8 +2181,9 @@ private void handlePostListPaginatedUsers(String filter, int limit, int offset,
 
-        if (USERNAME_CLAIM_URI.equalsIgnoreCase(claim) || SCIM_USERNAME_CLAIM_URI.equalsIgnoreCase(claim) ||
-                SCIM2_USERNAME_CLAIM_URI.equalsIgnoreCase(claim)) {
+        if (!isUniqueUserIdEnabled() && (USERNAME_CLAIM_URI.equalsIgnoreCase(claim)
```
why do we need to check !isUniqueUserIdEnabled()  here?

IsuraD commented on core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/common/AbstractUserStoreManager.java:
```diff
@@ -2766,14 +3759,16 @@ public final void addUser(String userName, Object credential, String[] roleList,
 
-        UserStore userStore = getUserStore(userName);
+        // Get the user store that this user should be added from the domain name that is appended to the username.
+        UserStore userStore = getUserStoreWithID(userName);
```
Why do we have to create new method named getUserStoreWithID

ashensw committed:
Fix code review changes

ashensw committed:
Refactor the code

JKAUSHALYA committed:
Change new APIs to use lists instead of arrays.

ashensw committed:
Refactor the code

JKAUSHALYA committed:
Change the unique id logic to call the new listeners.

JKAUSHALYA committed:
Change the unique id logic to call the new listeners. (#2365)

ashensw committed:
Add new getUserListOfRoleWithID API

ashensw committed:
Fix list users method

JKAUSHALYA committed:
Fix get user claims method.

ashensw committed:
Refactor the code

JKAUSHALYA committed:
Change the logic to get the unique id from internal method.

JKAUSHALYA committed:
Fix issue where unique id is retrieved from unique id manager.

JKAUSHALYA committed:
Change the logic to get the unique id from internal method. (#2370)

* Change the logic to get the unique id from internal method.

* Fix issue where unique id is retrieved from unique id manager.

JKAUSHALYA committed:
Change the logic to get the unique id from internal method. (#2371)

* Change the logic to get the unique id from internal method.

* Fix issue where unique id is retrieved from unique id manager.

* Remove unique id manger call from the get user id by username method.

JKAUSHALYA committed:
Add mechanism to retrieve list of unique IDs.

Commit:
Fix test failures.

ashensw committed:
Refactor the code

ashensw committed:
Fix test failures

ashensw committed:
Fix test failures

ashensw committed:
Fix getUserList method (#2379)

Commit:
Fix issues and test failures

Commit:
Fix issue in getRolesListOfUser method

Commit:
Fixing test failures

Commit:
Fixing test failures

Commit:
Fix issues idenfied though unit tests.

IsuraD committed:
Fix issues idenfied though unit tests. (#2386)

ashensw committed:
Fix issues (#2385)

ashensw committed:
Fix doCheckExistingUserWithID method

ashensw committed:
Fix profile loading issue (#2388)

JKAUSHALYA committed:
Add doc comments to the interface.

Commit:
Fix failed test cases.

ashensw committed:
Refactor the getTenantDomain mehod (#2390)

ashensw committed:
Refactor getUser method (#2392)

JKAUSHALYA committed:
Add method to get full qualified username.

JKAUSHALYA committed:
Fix issue where domain separator appended when userstore domain is empty

ashensw committed:
Refactor the code (#2395)

JKAUSHALYA committed:
Add equals and hash code methods to user object.

JKAUSHALYA committed:
Change method to "getUserIDFromUserName" from "getUserIDByUserName".

Commit:
Fixing issue while adding a tenant.

JKAUSHALYA committed:
Improve code quality by restructuring.

Commit:
Fix issue while adding a tenant.

Commit:
Fix issue

JKAUSHALYA committed:
Fix invalid method name in call secure method. (#2402)

JKAUSHALYA committed:
Fix invalid method arguments in call secure method.

ashensw committed:
Add UniqueID UserStoreManagers (#2403)

This will introduce UniqueIDReadOnlyLDAPUserStoreManager and UniqueIDReadWriteLDAPUserStoreManager.

ashensw committed:
Add UserIDAttribute property to user-mgt.xml (#2405)

Commit:
Fix taking isEnableUniqueIdentifier property in correct manner

ashensw committed:
Refactor the code

JKAUSHALYA committed:
Fix issue where user id used instead of username.

ashensw committed:
Merge pull request #2408 from JKAUSHALYA/feature-unique-user-id

Fix issue where user id used instead of username.

ashensw committed:
Refactor the code

Commit:
Fixing issues in secondary user stores.

Commit:
Fixing issues in secondary user stores.

ashensw committed:
Merge pull request #2409 from IsuraD/feature2

Fixing issues in secondary user stores.

ashensw committed:
Add UniqueIDActiveDirectoryUserStoreManager (#2410)

Commit:
Fixing integration test failures

ashensw committed:
Merge pull request #2411 from IsuraD/feature2

Fixing integration test failures

ashensw committed:
Remove UniqueUserID property from carbon.xml

ashensw committed:
Merge pull request #2412 from ashensw/feature-unique-user-id

Remove UniqueUserID property from carbon.xml

ashensw committed:
Refactor the code (#2413)

JKAUSHALYA committed:
Change the method to read to username from user attributes table. (#2414)

Commit:
Fix secondary domain appenidng issue in listUsers method.

IsuraD committed:
Fix secondary domain appenidng issue in listUsers method. (#2415)

ashensw committed:
Fix test failures (#2416)

Commit:
Merge branch 'feature-unique-user-id' of https://github.com/wso2/carbon4-kernel into feature2

ashensw committed:
Update UniqueIDReadOnlyLDAPUserStoreManager.java

ashensw committed:
Merge pull request #2418 from ashensw/feature-unique-user-id

Update UniqueIDReadOnlyLDAPUserStoreManager.java

Commit:
Update licence header

ashensw committed:
Merge pull request #2421 from IsuraD/feature2

Feature2

ashensw committed:
Refactor the code (#2422)

ashensw committed:
Add UserIDEnabled property to the userstore configs. (#2423)

ashensw committed:
Fix User ID Enabled property loading issue

JKAUSHALYA committed:
Merge pull request #2424 from ashensw/feature-unique-user-id

Fix User ID Enabled property loading issue

ashensw committed:
Refactor the code (#2426)

Commit:
FIx identified issues.

ashensw committed:
Merge pull request #2427 from IsuraD/feature2

FIx identified issues.

ashensw committed:
Merge branch '4.6.x' of https://github.com/wso2/carbon-kernel into feature-unique-user-id

ashensw committed:
Merge pull request #2428 from ashensw/feature-unique-user-id

Merge changes from 4.6.x branch
